### ISSUE TITLE
Splitting generated code in a monorepo

### ISSUE THREAD

**Issue by fubhy** [bufbuild/protobuf-es#481]
It'd be great if there was a way to reference generated code between workspace packages in a monorepo instead of having them re-generated in each individual package.

For instance, I have a setup like this:

```
./apps/my-frontend/
./apps/my-frontend/proto
./apps/my-backend/
./apps/my-backend/proto
./packages/shared-package/
./packages/shared-package/proto/
./buf.work.yaml
```

All these packages & apps also rely on protos from external dependencies.

Ideally, I'd want to external dependency to be generated into `shared-package` and then imported from the others. However, currently there seems to be no way to achieve that and the same code for that external package is generated into each app and package over and over again.

Am I missing sth.? :-)

**Comment by fubhy** [comment]
Maybe sth. like this could be accomplished with a custom `google.protobuf.FileOptions` that lets you specify a mapping of proto import path to es module import path and if a record exists for an import then `protoc-gen-es` would use `import ... from "other module"` in the generated  code instead of generating that part itself?

**Comment by smaye81** [comment]
Hi @fubhy. We actually do something similar with the BSR. Would it be possible for you to generate your code into its own workspace package and use the `rewrite_imports` option to change the imports to this generated package? 

For example, the usage is an option to the plugin:

`opt: rewrite_imports=@bufbuild/protobuf:my/custom/path/protobuf.ts`

**Comment by fubhy** [comment]
Hi @smaye81 . I'm afraid that's not sufficient because it'd result in imports like:

```
import { Foo } from "@myscope/mypackage/foo/bar.js";
```

But what I'd like to get is:

```
import { Foo } from "@myscope/mypackage";
```

Maybe we can allow regex with substitutions in the patterns in the opts?

**Comment by fubhy** [comment]
I can make it work with glob exports. Not sure how well they are supported across the ecosystem though ...

```

  "exports": {
    ".": {
      "types": "./dist/types/index.d.ts",
      "import": "./dist/esm/index.js",
      "default": "./dist/cjs/index.js"
    },
    "./proto/*.js": {
      "types": "./dist/types/generated/*.d.ts",
      "import": "./dist/esm/generated/*.js",
      "default": "./dist/cjs/generated/*.js"
    },
    "./package.json": "./package.json"
  },
  "typesVersions": {
    "*": {
      "proto/*": ["./dist/types/generated/*.d.ts"]
    }
  },
  ```

This works so far but I'd still prefer to be able to control the full path substitution.

**Comment by smaye81** [comment]
Can you provide a more reproducible example of what you're trying to accomplish. I'm pretty sure we do something quite similar in the BSR so off the top of my head, I feel like this _should_ work, but it's hard to say for sure.

**Comment by fubhy** [comment]
> Can you provide a more reproducible example of what you're trying to accomplish. I'm pretty sure we do something quite similar in the BSR so off the top of my head, I feel like this _should_ work, but it's hard to say for sure.

Sure. Does this work for you? https://github.com/fubhy/substreams-es/blob/main/packages/substreams-proxy/buf.gen.yaml#L6-L12

That's the codebase in question.

I'm exporting the generated code from here: https://github.com/fubhy/substreams-es/blob/main/packages/substreams/package.json#L42-L46

This is my workaround because ideally I'd like to just export them from the default entrypoint (@fubhy/substreams) and have the import paths subsituted in @fubhy/substreams-proxy to import straight from that default entrypoint. But unless I'm missing sth. with `rewrite_imports`, it only rewrites the prefix but does not in fact let you rewrite in a way that the trailing path is dropped entirely.

**Comment by fubhy** [comment]
Sorry yeah I migrated the repo to an org and renamed the packages.

**Comment by smaye81** [comment]
It does indeed do that replacement for relative paths, but if you don't use a relative path, it should replace the entire thing.

For example:

```typescript
const Bar = f.import("Bar", "foo/bar_pb.js")
```
with the `rewrite_import` option of `"rewrite_imports=foo/**/*_pb.js:@scope/pkg"`

should result in an import of:

```typescript
import { Bar } from "@scope/pkg";
```

**Comment by fubhy** [comment]
Ah, good to know. But what about alternative exports entrypoints? :-P

**Comment by smaye81** [comment]
It could be more flexible, yeah, but this is actually an undocumented secret menu item. It was created to assist us with something in the BSR and isn't 'officially' supported as an option. I think it was made about as flexible as we needed it to be, but Hyrum's Law took over and the rest is history.

Hopefully you can get the full replacement working though using non-relative imports. If not, let me know what you're trying and I'll see if I can help.

**Comment by fubhy** [comment]
> Hyrum's Law took over and the rest is history.

Heh ;-)

> Hopefully you can get the full replacement working though using non-relative imports. If not, let me know what you're trying and I'll see if I can help.

I'm fine for now. Don't want you to waste time on that ;-). I'm going with the wildcard export solution that I posted earlier for now. If you are open to making this a full feature & more versatile in the future (e.g. with some basic regex capture group replacement or sth.), I'm happy to work on that.

**Comment by smaye81** [comment]
Thanks @fubhy.  We will definitely keep you posted if anything changes with this option.

Going to close this now, but appreciate the understanding.

**Comment by timostamm** [comment]
We've released [v2.15.0](https://github.com/bufbuild/protobuf-es/releases/tag/v2.15.0) with the new [plugin option map_imports](https://protobufes.com/reference/plugin-options/#map_imports). It's designed for use cases like this, but it will write imports with the full generated file path, e.g. `@scope/pkg/foo_bar.js`, so the package needs to use subpath exports. That's necessary to avoid name collisions.

Here's an example with `@bufbuild/protovalidate`: https://github.com/bufbuild/protovalidate-es/blob/79f38d71e0418c408e198cf34f5ba890eb1d891c/packages/example/buf.gen.yaml
